File: src/ContactManager.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;

public class ContactManager {
    private final ContactLinkedList contactList;
    private final HashMap<Integer, Contact> contactMap;

    // Always kept sorted by name, then ID.
    private final ArrayList<Contact> sortedContacts;

    private final Comparator<Contact> nameComparator =
            Comparator.comparing(Contact::getName, String.CASE_INSENSITIVE_ORDER)
                      .thenComparingInt(Contact::getId);

    public ContactManager() {
        contactList = new ContactLinkedList();
        contactMap = new HashMap<>();
        sortedContacts = new ArrayList<>();
    }

    public boolean addContact(Contact contact) {
        if (contactMap.containsKey(contact.getId())) {
            return false;
        }

        contactList.add(contact);
        contactMap.put(contact.getId(), contact);

        // Insert into the sorted collection.
        int position = findInsertionPosition(contact);
        sortedContacts.add(position, contact);

        return true;
    }

    public boolean deleteContact(int id) {
        Contact contact = contactMap.get(id);

        if (contact == null) {
            return false;
        }

        contactList.deleteById(id);
        contactMap.remove(id);

        // Remove the same object from the sorted collection.
        sortedContacts.remove(contact);

        return true;
    }

    public Contact searchById(int id) {
        return contactMap.get(id);
    }

    /*
     * Finds all contacts with the requested name using binary search.
     *
     * First binary-search for one matching contact, then expand left/right
     * because duplicate names are allowed.
     */
    public List<Contact> searchByName(String name) {
        ArrayList<Contact> results = new ArrayList<>();

        if (sortedContacts.isEmpty()) {
            return results;
        }

        int index = binarySearchName(name);

        if (index == -1) {
            return results;
        }

        int left = index;
        while (left >= 0 &&
                sortedContacts.get(left).getName().equalsIgnoreCase(name)) {
            left--;
        }

        int right = index;
        while (right < sortedContacts.size() &&
                sortedContacts.get(right).getName().equalsIgnoreCase(name)) {
            right++;
        }

        for (int i = left + 1; i < right; i++) {
            results.add(sortedContacts.get(i));
        }

        return results;
    }

    // Binary search by name.
    private int binarySearchName(String target) {
        int low = 0;
        int high = sortedContacts.size() - 1;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            String currentName = sortedContacts.get(mid).getName();

            int comparison = currentName.compareToIgnoreCase(target);

            if (comparison == 0) {
                return mid;
            } else if (comparison < 0) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return -1;
    }

    private int findInsertionPosition(Contact contact) {
        int low = 0;
        int high = sortedContacts.size();

        while (low < high) {
            int mid = low + (high - low) / 2;

            if (nameComparator.compare(sortedContacts.get(mid), contact) < 0) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }

        return low;
    }

    public void displayContacts() {
        contactList.displayRecursive();
    }

    public int countContacts() {
        return contactList.countRecursive();
    }
}

File: src/contact_manager.py
"""Smart Contact Search System implemented with Python data structures."""

from dataclasses import dataclass


@dataclass(frozen=True)
class Contact:
    id: int
    name: str
    phone: str
    email: str

    def __post_init__(self):
        if not isinstance(self.id, int):
            raise TypeError("id must be an integer")
        for field_name in ("name", "phone", "email"):
            if not isinstance(getattr(self, field_name), str):
                raise TypeError(f"{field_name} must be a string")
            if not getattr(self, field_name).strip():
                raise ValueError(f"{field_name} cannot be empty")


@dataclass
class Node:
    data: Contact
    next: "Node | None" = None


class ContactLinkedList:
    def __init__(self):
        self.head = None
        self._size = 0

    def add(self, contact):
        new_node = Node(contact)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next is not None:
                current = current.next
            current.next = new_node
        self._size += 1

    def delete_by_id(self, contact_id):
        if self.head is None:
            return False
        if self.head.data.id == contact_id:
            self.head = self.head.next
            self._size -= 1
            return True

        current = self.head
        while current.next is not None:
            if current.next.data.id == contact_id:
                current.next = current.next.next
                self._size -= 1
                return True
            current = current.next
        return False

    # Base case: no node means there are no contacts left to collect.
    # Recursive step: collect this node, then continue with the next node.
    def _to_list_recursive(self, current):
        if current is None:
            return []
        return [current.data] + self._to_list_recursive(current.next)

    def display_recursive(self):
        """Return contacts in linked-list order using recursive traversal."""
        # Start at the first node; the helper follows every next reference.
        return self._to_list_recursive(self.head)

    def count_recursive(self):
        # An empty list has zero contacts.
        if self.head is None:
            return 0

        def count(current):
            # Base case: the end of the list contributes zero.
            if current is None:
                return 0
            # Count this node, then recursively count the remaining nodes.
            return 1 + count(current.next)

        return count(self.head)

    def __len__(self):
        # Return the maintained size without traversing the list.
        return self._size


class ContactManager:
    def __init__(self):
        self.contact_list = ContactLinkedList()
        self._contacts_by_id = {}
        self._sorted_contacts = []

    def add_contact(self, contact):
        if contact.id in self._contacts_by_id:
            return False
        self.contact_list.add(contact)
        self._contacts_by_id[contact.id] = contact
        self._sorted_contacts.append(contact)
        self._sorted_contacts.sort(key=lambda item: (item.name.casefold(), item.id))
        return True

    def delete_contact(self, contact_id):
        contact = self._contacts_by_id.get(contact_id)
        if contact is None:
            return False
        self.contact_list.delete_by_id(contact_id)
        del self._contacts_by_id[contact_id]
        self._sorted_contacts.remove(contact)
        return True

    def search_by_id(self, contact_id):
        return self._contacts_by_id.get(contact_id)

    def search_by_name(self, name):
        target = name.casefold()
        low, high = 0, len(self._sorted_contacts) - 1
        match_index = -1
        while low <= high:
            middle = (low + high) // 2
            current_name = self._sorted_contacts[middle].name.casefold()
            if current_name == target:
                match_index = middle
                break
            if current_name < target:
                low = middle + 1
            else:
                high = middle - 1

        if match_index == -1:
            return []
        first = match_index
        while first > 0 and self._sorted_contacts[first - 1].name.casefold() == target:
            first -= 1
        last = match_index
        while last + 1 < len(self._sorted_contacts) and self._sorted_contacts[last + 1].name.casefold() == target:
            last += 1
        return self._sorted_contacts[first:last + 1]

    def display_contacts(self):
        return self.contact_list.display_recursive()

    def count_contacts(self):
        return self.contact_list.count_recursive()
